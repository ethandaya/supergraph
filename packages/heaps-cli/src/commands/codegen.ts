import { CAC } from "cac";
import { loadConfig } from "../utils/load";
import {
  EntityGenerator,
  EventGenerator,
  ModelGenerator,
} from "@heaps/generators";
import fs from "fs";
import { watch } from "chokidar";

export type CodegenOptions = {
  watch: boolean;
  pathToSchema: string;
  // TODO - design of store path needs a lot of work
  storeImportPath: string;
  pathToConfig: string;
  outputDir: string;
};
export function buildArtifacts(options: Omit<CodegenOptions, "watch">) {
  const config = loadConfig(options);
  const { pathToSchema, outputDir } = options;
  const modelGenerator = new ModelGenerator({
    schemaPath: pathToSchema,
    outputPath: outputDir + "/models.ts",
  });
  const entityGenerator = new EntityGenerator({
    isAsync: false,
    schemaPath: pathToSchema,
    outputPath: outputDir + "/schema.ts",
    storeImportPath: options.storeImportPath,
    modelImportPath: "./models",
  });
  for (const idx in config.sources) {
    const source = config.sources[idx];
    fs.mkdirSync(outputDir + `/${source.name}`, { recursive: true });
    const abi = JSON.parse(fs.readFileSync(source.abi, "utf-8"));
    const eventGenerator = new EventGenerator({
      abi,
      outputPath: outputDir + `/${source.name}/${source.name}.ts`,
    });
    eventGenerator.generate(true);
  }
  modelGenerator.generate(true);
  entityGenerator.generate(true);
}

export function codegen(options: CodegenOptions) {
  buildArtifacts(options);
  if (options.watch) {
    console.log("Watching for Changes...");
    watch(options.pathToSchema).on("change", async () => {
      console.log("Change detected, regenerating...");
      try {
        await codegen(options);
      } catch (e) {
        console.error("Error while regenerating");
        console.error(e);
      }
    });
  }
}

export function registerCodegenCommands(cli: CAC) {
  cli
    .command("codegen")
    .option("--watch", "Watch mode")
    .option("--pathToSchema", "Path to schema definition", {
      default: "./schema.graphql",
    })
    .option("--storeImportPath", "Path to store definition", {
      default: "../../store",
    })
    .option("--pathToConfig", "Path to supergraph config", {
      default: "./supergraph.json",
    })
    .option("--outputDir", "Directory to output artifacts too", {
      default: "./src/types",
    })
    .action(codegen);
  return cli;
}
