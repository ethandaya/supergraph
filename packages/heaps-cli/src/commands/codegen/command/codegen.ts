import { EntityGenerator, ModelGenerator } from "@heaps/generators";

type CodegenOptions = {
  watch: boolean;
  pathToSchema: string;
  outputDir: string;
};
export async function codegen(options: CodegenOptions) {
  const { pathToSchema, outputDir } = options;
  const modelGenerator = new ModelGenerator({
    schemaPath: pathToSchema,
    outputPath: outputDir + "/models.ts",
  });
  const entityGenerator = new EntityGenerator({
    isAsync: false,
    schemaPath: pathToSchema,
    outputPath: outputDir + "/schema.ts",
    storeImportPath: "./store",
    modelImportPath: outputDir + "./models",
  });
  await modelGenerator.generate(true);
  await entityGenerator.generate(true);
}
