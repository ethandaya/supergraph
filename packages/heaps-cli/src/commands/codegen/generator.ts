import {
  MethodDeclarationStructure,
  OptionalKind,
  Project,
  SourceFile,
} from "ts-morph";
import { z } from "zod";
import { makeGetterStatements, makeSetterStatements } from "./statements";

type GenerateOptions = {
  shouldSave: boolean;
};

type EntityGeneratorOptions = {
  models: { [key: string]: z.AnyZodObject };
  outputPath: string;
};

export class EntityGenerator {
  private project: Project;
  public targetFile: SourceFile;
  private readonly models: {
    [key: string]: z.AnyZodObject;
  } = {};

  constructor(options: EntityGeneratorOptions) {
    this.project = new Project();
    this.targetFile = this.project.createSourceFile(
      options.outputPath,
      undefined,
      {
        overwrite: true,
      }
    );
    this.models = options.models;
  }

  public generateImports() {
    this.targetFile.addImportDeclaration({
      namedImports: ["z"],
      moduleSpecifier: "zod",
    });
    this.targetFile.addImportDeclaration({
      namedImports: ["Entity", "KeyAccessError", "SQLiteStore as Store"],
      moduleSpecifier: "@heaps/engine",
    });
    this.targetFile.addImportDeclaration({
      namedImports: [...Object.keys(this.models)],
      moduleSpecifier: "../models",
    });
  }

  public generateTypeForSchema(name: string) {
    this.targetFile.addTypeAlias({
      name: name + "Model",
      type: `z.infer<typeof ${name}Schema>`,
    });
  }

  public generateClassForSchema(name: string, schema: z.ZodObject<any>) {
    const methods: OptionalKind<MethodDeclarationStructure>[] = [];

    methods.push({
      name: "load",
      isStatic: true,
      parameters: [
        {
          name: "id",
          type: "string",
        },
      ],
      returnType: `${name} | null`,
      statements: [
        `const data = store.get<${name}Model>("${name.toLowerCase()}", id);`,
        `if (!data) {`,
        `   return new ${name}(id);`,
        `}`,
        `return new ${name}(id, data);`,
      ],
    });

    for (const key in schema.shape) {
      const type = `${name}Model["${key}"]`;
      const isNullable = schema.shape[key].isNullable();
      methods.push({
        name: `get ${key}`,
        returnType: type,
        statements: makeGetterStatements({
          key,
          name,
          isNullable,
        }),
      });
      methods.push({
        name: `set ${key}`,
        parameters: [{ name: "value", type }],
        statements: makeSetterStatements(key),
      });
    }

    this.targetFile.addClass({
      name,
      isExported: true,
      extends: `Entity<${name}Model, typeof ${name}Schema>`,
      ctors: [
        {
          parameters: [
            {
              name: "id",
              type: "string",
            },
            {
              name: "data",
              type: `${name}Model`,
              hasQuestionToken: true,
            },
          ],
          statements: [
            `super(id, ${name}Schema, store)`,
            `this.data = { id, ...data } || {};`,
          ],
        },
      ],
      methods,
    });
  }

  public generateDefinitionsForModel(name: string, schema: z.ZodObject<any>) {
    this.generateTypeForSchema(name);
    this.generateClassForSchema(name, schema);
  }

  public generateGlobalDefinitions() {
    this.targetFile.addTypeAlias({
      name: "SchemaNames",
      type: Object.keys(this.models)
        .map((key) => `"${key.toLowerCase().replace("schema", "")}"`)
        .join(" | "),
    });
  }

  public generateStore() {
    this.targetFile.addStatements([
      `export const store = new Store<SchemaNames>("", {`,
      ...Object.keys(this.models).map(
        (key) => `  ${key.replace("Schema", "").toLowerCase()}: ${key},`
      ),
      `});`,
    ]);
  }

  public generate(options: GenerateOptions = { shouldSave: true }) {
    this.generateImports();
    this.generateGlobalDefinitions();
    this.generateStore();
    for (const name in this.models) {
      this.generateDefinitionsForModel(
        name.replace("Schema", ""),
        this.models[name]
      );
    }
    this.targetFile.prependWhitespace("\n\n");
    this.targetFile.insertText(
      0,
      "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY."
    );
    if (options.shouldSave) {
      this.save();
    }
  }

  public save() {
    this.targetFile.saveSync();
  }
}
