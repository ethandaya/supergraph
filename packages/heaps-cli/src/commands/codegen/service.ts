import * as swc from "@swc/core";
import * as path from "path";
import * as fs from "fs";
import { getFullPath } from "@heaps/common/src";

// TODO - this is bad, fix later
export function transformTS(path: string) {
  const module = swc.transformFileSync(path, {
    jsc: {
      target: "es2022",
      parser: {
        syntax: "typescript",
      },
    },
    module: {
      type: "nodenext",
    },
  });
  return module.code;
}

export function prepBuildDir() {
  const buildDir = getFullPath(".heaps");
  if (!fs.existsSync(buildDir)) {
    fs.mkdirSync(buildDir);
  }
}

export function build(pathToSource: string): string {
  prepBuildDir();
  const buildPath = path.join(
    ".heaps",
    path.basename(pathToSource, ".ts") + ".js"
  );
  const code = transformTS(pathToSource);
  fs.writeFileSync(buildPath, code);
  return buildPath;
}
