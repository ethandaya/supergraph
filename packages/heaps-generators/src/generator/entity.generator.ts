import {
  MethodDeclarationStructure,
  OptionalKind,
  Project,
  SourceFile,
} from "ts-morph";
import * as fs from "fs";
import { format } from "prettier";
import { isDerivedField, isObjectTypeDefinition } from "../common";
import { DocumentNode, Kind, parse } from "graphql/language";
import { FieldDefinitionNode } from "graphql/index";

type EntityGeneratorOptions = {
  isAsync: boolean;
  outputPath: string;
  storeImportPath: string;
  modelImportPath: string;
  schemaPath: string;
};

type GetterStatementOptions = {
  key: string;
  name: string;
  isNullable: boolean;
};

export function makeGetterStatements({
  key,
  name,
  isNullable,
}: GetterStatementOptions) {
  return [
    `const value = this.get("${key}")`,
    `if (typeof value === "undefined"${
      isNullable ? " && value !== null" : ""
    }) {`,
    `  throw new KeyAccessError<${name}>("${key}")`,
    `}`,
    `return value`,
  ];
}

export function makeSetterStatements(key: string) {
  return `this.set("${key}", value);`;
}

export class EntityGenerator {
  private project: Project;
  public targetFile: SourceFile;
  public schema: DocumentNode;

  constructor(private readonly options: EntityGeneratorOptions) {
    this.project = new Project();
    this.targetFile = this.project.createSourceFile(
      options.outputPath,
      undefined,
      { overwrite: true }
    );
    this.schema = this.loadSchema(options.schemaPath);
  }

  private loadSchema(path: string) {
    const document = fs.readFileSync(path, "utf-8");
    return parse(document);
  }

  get entities() {
    return this.schema.definitions.filter(isObjectTypeDefinition);
  }

  public generateImports() {
    const imports = [
      {
        namedImports: ["z"],
        moduleSpecifier: "zod",
      },
      {
        namedImports: ["CrudEntity", "KeyAccessError"],
        moduleSpecifier: "@heaps/engine",
      },
      {
        namedImports: [...this.entities.map((e) => e.name.value + "Schema")],
        moduleSpecifier: this.options.modelImportPath,
      },
      {
        namedImports: ["store"],
        moduleSpecifier: this.options.storeImportPath,
      },
    ];

    this.targetFile.addImportDeclarations(imports);
  }

  public generateTypeForModel(name: string) {
    this.targetFile.addTypeAlias({
      name: name + "Model",
      type: `z.infer<typeof ${name}Schema>`,
    });
  }

  public generateEntityForModel(
    name: string,
    fields: readonly FieldDefinitionNode[]
  ) {
    let methods: OptionalKind<MethodDeclarationStructure>[] = [];

    const fieldsToGenerate = fields.filter(
      (field) => !isDerivedField(field.directives)
    );

    for (const idx in fieldsToGenerate) {
      const field = fieldsToGenerate[idx];
      const key = field.name.value;
      const type = `${name}Model["${key}"]`;
      const isNullable = field.type.kind === Kind.NON_NULL_TYPE;
      methods.push({
        name: `get ${key}`,
        returnType: type,
        statements: makeGetterStatements({
          key,
          name,
          isNullable,
        }),
      });
      methods.push({
        name: `set ${key}`,
        parameters: [{ name: "value", type }],
        statements: makeSetterStatements(key),
      });
    }

    methods.push({
      name: "load",
      isAsync: this.options.isAsync,
      isStatic: true,
      parameters: [
        {
          name: "id",
          type: "string",
        },
      ],
      returnType: this.options.isAsync
        ? `Promise<${name} | null>`
        : `${name} | null`,
      statements: [
        `const data =${
          this.options.isAsync ? " await" : ""
        } store.get("${name.toLowerCase()}", id);`,
        `if (!data) {`,
        `   return null`,
        `}`,
        `return new ${name}(id, data);`,
      ],
    });

    methods.push({
      name: "save",
      isAsync: this.options.isAsync,
      statements: [
        `const dto = this._schema.extend({`,
        `updatedAt: z.bigint().optional(),`,
        `createdAt: z.bigint().optional(),`,
        `}).parse({ id: this._id, ...this._data });`,
        `this._data = ${
          this.options.isAsync ? "await" : ""
        } store.set("${name.toLowerCase()}", this.id, dto);`,
        `return this._data;`,
      ],
    });

    this.targetFile.addClass({
      name,
      isExported: true,
      extends: `CrudEntity<"${name}", ${name}Model, typeof ${name}Schema>`,
      ctors: [
        {
          parameters: [
            {
              name: "id",
              type: "string",
            },
            {
              name: "data",
              type: `${name}Model`,
              hasQuestionToken: true,
            },
          ],
          statements: [
            `super(id, "${name}", ${name}Schema)`,
            `this._data = { id, ...data } || { id };`,
          ],
        },
      ],
      methods,
    });
  }

  public generateDefinitionsForModel(
    name: string,
    fields: readonly FieldDefinitionNode[]
  ) {
    this.generateTypeForModel(name);
    this.generateEntityForModel(name, fields);
  }

  public generateEntities() {
    for (const en in this.entities) {
      const entity = this.entities[en];
      this.generateDefinitionsForModel(entity.name.value, entity.fields || []);
    }
  }

  public generate(save = false) {
    this.generateImports();
    this.generateEntities();
    if (save) {
      const formatted = format(this.targetFile.getText(), {
        parser: "typescript",
      });
      const header = `// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.`;
      fs.writeFileSync(this.options.outputPath, `${header}\n\n${formatted}`);
    }
  }
}
