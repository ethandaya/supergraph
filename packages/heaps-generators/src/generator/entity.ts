import { Project, SourceFile } from "ts-morph";
import * as fs from "fs";
import { format } from "prettier";

type EntityGeneratorOptions = {
  isAsync: boolean;
  outputPath: string;
  storePath: string;
};

export class EntityGenerator {
  private project: Project;
  public targetFile: SourceFile;

  constructor(private readonly options: EntityGeneratorOptions) {
    this.project = new Project();
    this.targetFile = this.project.createSourceFile(
      options.outputPath,
      undefined,
      { overwrite: true }
    );
  }

  public generateImports() {
    const imports = [
      {
        namedImports: ["z"],
        moduleSpecifier: "zod",
      },
      {
        namedImports: [
          `${this.options.isAsync ? "Async" : "Sync"}CrudEntity`,
          "KeyAccessError",
        ],
        moduleSpecifier: "@heaps/engine",
      },
      {
        namedImports: ["store"],
        moduleSpecifier: this.options.storePath,
      },
    ];

    this.targetFile.addImportDeclarations(imports);
  }

  public generate(save = false) {
    this.generateImports();
    if (save) {
      const formatted = format(this.targetFile.getText(), {
        parser: "typescript",
      });
      fs.writeFileSync(this.options.outputPath, formatted);
    }
  }
}
