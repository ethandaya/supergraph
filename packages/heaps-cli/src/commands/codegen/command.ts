import { build } from "./service";
import { EntityGenerator } from "./generator";
import { getFullPath } from "@heaps/common/src";

type CodegenOptions = {
  pathToModels: string;
  outputPath: string;
};

export async function codegen(options: CodegenOptions) {
  let { pathToModels, outputPath } = options;
  // TODO - this is bad but works for now
  const buildPath = build(pathToModels);
  const models = require(buildPath);
  const generator = new EntityGenerator({
    models,
    outputPath: getFullPath(outputPath),
  });
  generator.generate();
}
