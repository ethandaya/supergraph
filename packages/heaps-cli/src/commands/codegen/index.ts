import { CAC } from "cac";
import { codegen } from "./command";

export function registerCodegenCommands(cli: CAC) {
  cli
    .command("codegen")
    .option("--watch", "Watch mode")
    .option("--pathToSchema", "Path to schema definition", {
      default: "./schema.graphql",
    })
    .option("--storeImportPath", "Path to store definition", {
      default: "../../store",
    })
    .option("--pathToConfig", "Path to supergraph config", {
      default: "./supergraph.json",
    })
    .option("--outputDir", "Directory to output artifacts too", {
      default: "./src/types",
    })
    .action(codegen);
  return cli;
}
