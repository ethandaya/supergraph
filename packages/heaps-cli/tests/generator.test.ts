import { EntityGenerator } from "../src/commands/codegen/generator";
import os from "os";
import { z } from "zod";
import * as fs from "fs";

const TestSchema = z.object({
  id: z.string(),
  name: z.string(),
  age: z.number(),
  address: z.string(),
});

describe("Generator", () => {
  let generator: EntityGenerator;
  let outputPath: string;

  beforeEach(() => {
    outputPath = os.tmpdir() + "/schema.ts";
    generator = new EntityGenerator({
      models: { test: TestSchema },
      outputPath,
    });
  });

  it("should generate an entity & model definition for a schema with a nullable field", () => {
    generator.generateDefinitionsForModel("Test", TestSchema);
    expect(generator.targetFile.getFullText()).toMatchSnapshot();
  });

  it("should generate a global type for all schemas", () => {
    generator.generateGlobalDefinitions();
    expect(generator.targetFile.getFullText()).toMatchSnapshot();
  });
  it("should generate the full type set", () => {
    generator.generate({
      shouldSave: true,
    });
    expect(generator.targetFile.getFullText()).toMatchSnapshot();
  });
  it("should generate the full type set with a custom output path", () => {
    generator.generate({
      shouldSave: true,
    });
    expect(generator.targetFile.getFullText()).toMatch(
      fs.readFileSync(outputPath, "utf-8")
    );
  });
});
