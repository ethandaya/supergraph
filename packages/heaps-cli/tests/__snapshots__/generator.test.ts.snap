// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator should generate a global type for all schemas 1`] = `
"type SchemaNames = "test";
"
`;

exports[`Generator should generate an entity & model definition for a schema with a nullable field 1`] = `
"type TestModel = z.infer<typeof TestSchema>;

export class Test extends Entity<TestModel, typeof TestSchema> {
    constructor(id: string, data?: TestModel) {
        super(id, TestSchema, store)
        this.data = { id, ...data } || {};
    }

    static load(id: string): Test | null {
        const data = store.get<TestModel>("test", id);
        if (!data) {
           return new Test(id);
        }

        return new Test(id, data);
    }

    get id(): TestModel["id"] {
        const value = this.get("id")
        if (typeof value === "undefined") {
          throw new KeyAccessError<Test>("id")
        }

        return value
    }

    set id(value: TestModel["id"]) {
        this.set("id", value);
    }

    get name(): TestModel["name"] {
        const value = this.get("name")
        if (typeof value === "undefined") {
          throw new KeyAccessError<Test>("name")
        }

        return value
    }

    set name(value: TestModel["name"]) {
        this.set("name", value);
    }

    get age(): TestModel["age"] {
        const value = this.get("age")
        if (typeof value === "undefined") {
          throw new KeyAccessError<Test>("age")
        }

        return value
    }

    set age(value: TestModel["age"]) {
        this.set("age", value);
    }

    get address(): TestModel["address"] {
        const value = this.get("address")
        if (typeof value === "undefined") {
          throw new KeyAccessError<Test>("address")
        }

        return value
    }

    set address(value: TestModel["address"]) {
        this.set("address", value);
    }
}
"
`;

exports[`Generator should generate the full type set 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

import { z } from "zod";
import { Entity, KeyAccessError, SQLiteStore as Store } from "@heaps/engine";
import { test } from "../models";

type SchemaNames = "test";
export const store = new Store<SchemaNames>("", {
  test: test,
});

type testModel = z.infer<typeof testSchema>;

export class test extends Entity<testModel, typeof testSchema> {
    constructor(id: string, data?: testModel) {
        super(id, testSchema, store)
        this.data = { id, ...data } || {};
    }

    static load(id: string): test | null {
        const data = store.get<testModel>("test", id);
        if (!data) {
           return new test(id);
        }

        return new test(id, data);
    }

    get id(): testModel["id"] {
        const value = this.get("id")
        if (typeof value === "undefined") {
          throw new KeyAccessError<test>("id")
        }

        return value
    }

    set id(value: testModel["id"]) {
        this.set("id", value);
    }

    get name(): testModel["name"] {
        const value = this.get("name")
        if (typeof value === "undefined") {
          throw new KeyAccessError<test>("name")
        }

        return value
    }

    set name(value: testModel["name"]) {
        this.set("name", value);
    }

    get age(): testModel["age"] {
        const value = this.get("age")
        if (typeof value === "undefined") {
          throw new KeyAccessError<test>("age")
        }

        return value
    }

    set age(value: testModel["age"]) {
        this.set("age", value);
    }

    get address(): testModel["address"] {
        const value = this.get("address")
        if (typeof value === "undefined") {
          throw new KeyAccessError<test>("address")
        }

        return value
    }

    set address(value: testModel["address"]) {
        this.set("address", value);
    }
}
"
`;
